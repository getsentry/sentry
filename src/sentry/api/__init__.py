from .analytics import *  # NOQA
from .client import ApiClient

client = ApiClient()
