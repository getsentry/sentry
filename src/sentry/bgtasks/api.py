from __future__ import absolute_import

import six
import time
import random
import logging
import threading
from contextlib import contextmanager

from django.conf import settings


logger = logging.getLogger("sentry.bgtasks")
tasks = {}


def bgtask(roles=None, interval=60):
    def decorator(f):
        return BgTask(callback=f, roles=roles, interval=interval)

    return decorator


class BgTask(object):
    def __init__(self, callback, roles=None, interval=60):
        self.callback = callback
        self.roles = roles or []
        self.interval = interval
        self.running = False

    @property
    def name(self):
        return "%s:%s" % (self.callback.__module__, self.callback.__name__)

    def run(self):
        if self.running:
            return
        self.running = True

        next_run = time.time() + self.interval * random.random()
        while self.running:
            now = time.time()
            if now >= next_run:
                try:
                    self.callback()
                except Exception:
                    logging.error("bgtask.failed", exc_info=True, extra=dict(task_name=self.name))
                next_run = now + self.interval
            time.sleep(1.0)

    def reconfigure(self, cfg):
        if "roles" in cfg:
            self.roles = cfg["roles"]
        if "interval" in cfg:
            self.interval = cfg["interval"]

    def spawn_daemon(self):
        if self.running:
            return
        logger.info("bgtask.spawn", extra=dict(task_name=self.name))
        t = threading.Thread(target=self.run)
        t.setDaemon(True)
        t.start()

    def stop(self):
        logger.info("bgtask.stop", extra=dict(task_name=self.name))
        self.running = False


def get_task(task_name):
    module, task_cls = task_name.split(":", 1)
    mod = __import__(module, None, None, [task_cls])
    return getattr(mod, task_cls)


def spawn_bgtasks(role):
    for import_name, cfg in six.iteritems(settings.BGTASKS):
        task = get_task(import_name)
        # This is already running
        if task.name in tasks:
            continue
        task.reconfigure(cfg)
        if role not in task.roles:
            continue
        task.spawn_daemon()
        tasks[task.name] = task


def shutdown_bgtasks():
    for task_name, task in list(six.iteritems(tasks)):
        task.stop()
        tasks.pop(task_name, None)


@contextmanager
def managed_bgtasks(role):
    spawn_bgtasks(role)
    try:
        yield
    finally:
        shutdown_bgtasks()
