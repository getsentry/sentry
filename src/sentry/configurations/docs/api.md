# Configurations API

Host: https://sentry.io/api/0

**Authors.**

@cmanallen

## Configuration [/projects/<organization_id_or_slug>/<project_id_or_slug>/configuration/]

### Get Configuration [GET]

Retrieve the project's configuration.

**Attributes**

| Column             | Type   | Description                     |
| ------------------ | ------ | ------------------------------- |
| id                 | number | Project id.                     |
| sample_rate        | number | A number value between 0 and 1. |
| traces_sample_rate | number | A number value between 0 and 1. |
| user_config        | any    | Arbitrary user supplied JSON.   |

- Response 200

  ```json
  {
    "data": {
      "id": 1,
      "sample_rate": 1.0,
      "traces_sample_rate": 0.5,
      "user_config": {
        "hello": "world"
      }
    }
  }
  ```

### Update Configuration [PATCH]

Update the project's configuration.

- Request

  ```json
  {
    "data": {
      "sample_rate": 0.2,
      "traces_sample_rate": 0.5,
      "user_config": ["hello", "world"]
    }
  }
  ```

- Response 201

  ```json
  {
    "data": {
      "id": 1,
      "sample_rate": 0.2,
      "traces_sample_rate": 0.5,
      "user_config": ["hello", "world"]
    }
  }
  ```

### Delete Configuration [DELETE]

Delete the project's configuration.

- Response 204
