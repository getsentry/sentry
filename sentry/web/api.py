"""
sentry.web.views
~~~~~~~~~~~~~~~~

:copyright: (c) 2010 by the Sentry Team, see AUTHORS for more details.
:license: BSD, see LICENSE for more details.
"""
import logging

from django.http import HttpResponse, HttpResponseBadRequest, \
  HttpResponseForbidden
from django.views.decorators.csrf import csrf_exempt
from django.views.decorators.http import require_http_methods

from sentry.conf import settings
from sentry.coreapi import project_from_auth_vars, project_from_api_key_and_id, \
  project_from_id, decode_and_decompress_data, safely_load_json_string, \
  validate_data, insert_data_to_database, APIError, APIUnauthorized, \
  extract_auth_vars, InvalidTimestamp
from sentry.models import Group, GroupBookmark, Project, View, ProjectDomain
from sentry.utils import json
from sentry.utils.http import is_same_domain, apply_access_control_headers
from sentry.web.decorators import has_access
from sentry.web.frontend.groups import _get_group_list
from sentry.web.helpers import render_to_response, render_to_string

logger = logging.getLogger('sentry.errors.coreapi')


@csrf_exempt
@require_http_methods(['POST', 'OPTIONS'])
def store(request):
    """
    The primary endpoint for storing new events.

    This will validate the client's authentication and data, and if
    successfull pass on the payload to the internal database handler.

    Authentication works in three flavors:

    1. Explicit signed requests

       These are implemented using the documented signed request protocol, and
       require an authentication header which is signed using with the project
       member's secret key.

    2. Explicit trusted requests

       Generally used for communications with client-side platforms (such as
       JavaScript in the browser), they require the GET variables public_key
       and project_id, as well as an HTTP_REFERER to be set from a trusted
       domain.

    3. Implicit trusted requests

       Used by the Sentry core, they are only available from same-domain requests
       and do not require any authentication information. They only require that
       the user be authenticated, and a project_id be sent in the GET variables.

    """
    try:
        if request.method == 'POST':
            auth_vars = extract_auth_vars(request)
            data = request.raw_post_data

            if auth_vars:
                server_version = auth_vars.get('sentry_version', '1.0')
                client = auth_vars.get('sentry_client')
            else:
                server_version = request.GET.get('version', '1.0')
                client = request.META.get('HTTP_USER_AGENT', request.GET.get('client'))

            if server_version not in ('1.0', '2.0'):
                raise APIError('Client/server version mismatch: Unsupported version: %r' % server_version)

            if server_version != '1.0' and not client:
                raise APIError('Client request error: Missing client version identifier.')

            referrer = request.META.get('HTTP_REFERER')

            if auth_vars:
                project = project_from_auth_vars(auth_vars, data)
            elif request.GET.get('api_key') and request.GET.get('project_id'):
                # public requests only need referrer validation for CSRF
                project = project_from_api_key_and_id(request.GET['api_key'], request.GET['project_id'])
                if not ProjectDomain.test(project, referrer):
                    raise APIUnauthorized()
            elif request.GET.get('project_id') and request.user.is_authenticated() and \
                 is_same_domain(request.build_absolute_uri(), referrer):
                # authenticated users are simply trusted to provide the right id
                project = project_from_id(request)
            else:
                raise APIUnauthorized()

            if not data.startswith('{'):
                data = decode_and_decompress_data(data)
            data = safely_load_json_string(data)

            try:
                validate_data(project, data)
            except InvalidTimestamp:
                # Log the error, remove the timestamp, and revalidate
                logger.error('Client %r passed an invalid value for timestamp %r' % (
                    data['timestamp'],
                    client or '<unknown client>',
                ))
                del data['timestamp']
                validate_data(project, data)

            insert_data_to_database(data)
    except APIError, error:
        response = HttpResponse(error.msg, status=error.http_status)
    else:
        response = HttpResponse('')
    return apply_access_control_headers(response)


@csrf_exempt
@has_access
def notification(request, project):
    return render_to_response('sentry/partial/_notification.html', request.GET)


@csrf_exempt
@has_access
def poll(request, project):
    from sentry.templatetags.sentry_helpers import as_bookmarks, handle_before_events

    offset = 0
    limit = settings.MESSAGES_PER_PAGE

    view_id = request.GET.get('view_id')
    if view_id:
        try:
            view = View.objects.get(pk=view_id)
        except View.DoesNotExist:
            return HttpResponseBadRequest()
    else:
        view = None

    filters, event_list = _get_group_list(
        request=request,
        project=project,
        view=view,
    )

    event_list = list(event_list[offset:limit])
    handle_before_events(request, event_list)

    data = [
        {
            'id': m.pk,
            'html': render_to_string('sentry/partial/_group.html', {
                'group': m,
                'request': request,
                'is_bookmarked': b,
            }).strip(),
            'title': m.message_top(),
            'message': m.error(),
            'level': m.get_level_display(),
            'logger': m.logger,
            'count': m.times_seen,
            'score': getattr(m, 'sort_value', None),
        } for m, b in as_bookmarks(event_list, request.user)]

    response = HttpResponse(json.dumps(data))
    response['Content-Type'] = 'application/json'
    return response


@csrf_exempt
@has_access
def resolve(request, project):
    gid = request.REQUEST.get('gid')
    if not gid:
        return HttpResponseForbidden()
    try:
        group = Group.objects.get(pk=gid)
    except Group.DoesNotExist:
        return HttpResponseForbidden()

    Group.objects.filter(pk=group.pk).update(status=1)
    group.status = 1

    data = [
        (m.pk, {
            'html': render_to_string('sentry/partial/_group.html', {
                'group': m,
                'request': request,
            }).strip(),
            'count': m.times_seen,
        }) for m in [group]]

    response = HttpResponse(json.dumps(data))
    response['Content-Type'] = 'application/json'
    return response


@csrf_exempt
@has_access
def bookmark(request, project):
    gid = request.REQUEST.get('gid')
    if not gid:
        return HttpResponseForbidden()

    if not request.user.is_authenticated():
        return HttpResponseForbidden()

    try:
        group = Group.objects.get(pk=gid)
    except Group.DoesNotExist:
        return HttpResponseForbidden()

    gb, created = GroupBookmark.objects.get_or_create(
        project=group.project,
        user=request.user,
        group=group,
    )
    if not created:
        gb.delete()

    response = HttpResponse(json.dumps({'bookmarked': created}))
    response['Content-Type'] = 'application/json'
    return response


@csrf_exempt
@has_access
def clear(request, project):
    view_id = request.GET.get('view_id')
    if view_id:
        try:
            view = View.objects.get(pk=view_id)
        except View.DoesNotExist:
            return HttpResponseBadRequest()
    else:
        view = None

    _, event_list = _get_group_list(
        request=request,
        project=project,
        view=view,
    )

    event_list.update(status=1)

    data = []
    response = HttpResponse(json.dumps(data))
    response['Content-Type'] = 'application/json'
    return response


@csrf_exempt
@has_access
def chart(request, project):
    gid = request.REQUEST.get('gid')
    days = int(request.REQUEST.get('days', '90'))

    if gid:
        try:
            group = Group.objects.get(pk=gid)
        except Group.DoesNotExist:
            return HttpResponseForbidden()

        data = Group.objects.get_chart_data(group, max_days=days)
    else:
        data = Project.objects.get_chart_data(project, max_days=days)

    response = HttpResponse(json.dumps(data))
    response['Content-Type'] = 'application/json'
    return response
