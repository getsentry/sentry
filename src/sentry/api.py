"""
sentry.api
~~~~~~~~~~

:copyright: (c) 2010-2012 by the Sentry Team, see AUTHORS for more details.
:license: BSD, see LICENSE for more details.
"""

import base64
from django.contrib.auth.models import User, AnonymousUser
from tastypie import fields
from tastypie.authentication import Authentication
from tastypie.authorization import Authorization
from tastypie.api import Api
from tastypie.resources import ModelResource
from sentry.models import Project, Team, TeamMember, Group, Event, ProjectKey
from sentry.web.helpers import get_project_list


class SentryAuthentication(Authentication):
    def is_authenticated(self, request, **kwargs):
        """
        Simple basic auth where username is the public key, and password
        is the secret key.
        """
        auth = request.META.get('HTTP_AUTHORIZATION')
        if not auth:
            return False

        if not auth.startswith('Basic '):
            return

        try:
            decoded = base64.b64decode(auth.split(' ', 1)[-1])
            public, secret = decoded.split(':', 1)
        except Exception:
            return False

        try:
            apikey = ProjectKey.objects.get(public_key=public, secret_key=secret)
        except ProjectKey.DoesNotExist:
            return False

        request.projects = [apikey.project] or get_project_list(apikey.user)
        if apikey.user:
            request.teams = Team.objects.get_for_user(apikey.user)
        else:
            request.teams = []
        request.user = apikey.user or AnonymousUser()
        request.apikey = apikey

        return True

    def get_identifier(self, request):
        # XXX: no idea what this is for
        return request.apikey.id


class TeamAuthorization(Authorization):
    def apply_limits(self, request, object_list):
        if request.projects:
            return object_list.filter(project__in=request.projects)
        elif request.teams:
            return object_list.filter(pk__in=[t.pk for t in request.teams])
        return object_list.none()


class TeamBasedAuthorization(Authorization):
    def apply_limits(self, request, object_list):
        if request.projects:
            return object_list.filter(team__project__in=request.projects)
        elif request.teams:
            return object_list.filter(team__in=request.teams)
        return object_list.none()


class ProjectAuthorization(Authorization):
    def apply_limits(self, request, object_list):
        if request.projects:
            return object_list.filter(pk__in=[p.id for p in request.projects])
        elif request.teams:
            return object_list.filter(team__in=request.teams)
        return object_list.none()


class ProjectBasedAuthorization(Authorization):
    def apply_limits(self, request, object_list):
        if request.projects:
            return object_list.filter(project__in=request.projects)
        elif request.teams:
            return object_list.filter(project__team__in=request.teams)
        return object_list.none()


class UserResource(ModelResource):
    username = fields.CharField('username', readonly=True)
    # name = fields.CharField('first_name', readonly=True)

    class Meta:
        queryset = User.objects.all()
        detail_uri_name = 'username'
        include_resource_uri = False
        resource_name = 'user'
        detail_allowed_methods = []
        list_allowed_methods = []
        fields = ['username']
        authentication = SentryAuthentication()


class TeamResource(ModelResource):
    slug = fields.CharField(attribute="slug", readonly=True)
    owner = fields.ToOneField(UserResource, 'owner', full=True, readonly=True)
    members = fields.ToManyField('sentry.api.TeamMemberResource', 'member_set', full=True)

    class Meta:
        queryset = Team.objects.all()
        detail_uri_name = 'slug'
        resource_name = 'team'
        detail_allowed_methods = ['get']
        list_allowed_methods = ['get']
        fields = ['id', 'name']
        authentication = SentryAuthentication()
        authorization = TeamAuthorization()


class TeamMemberResource(ModelResource):
    user = fields.ToOneField(UserResource, 'user', full=True)
    # access = fields.CharField('type')

    class Meta:
        queryset = TeamMember.objects.all()
        include_resource_uri = False
        detail_allowed_methods = ['get']
        list_allowed_methods = []
        fields = ['user']
        authentication = SentryAuthentication()
        authorization = TeamBasedAuthorization()

    # def dehydrate_access(self, bundle):
    #     return bundle.obj.get_type_display()


class ProjectResource(ModelResource):
    slug = fields.CharField(attribute="slug", readonly=True)
    owner = fields.ToOneField(UserResource, 'owner', full=True, readonly=True)
    team = fields.ToOneField(TeamResource, 'team', full=True)

    class Meta:
        queryset = Project.objects.filter(status=0)
        detail_uri_name = 'slug'
        resource_name = 'project'
        fields = ['id', 'name']
        allowed_methods = ['get']
        authentication = SentryAuthentication()
        authorization = ProjectAuthorization()


class GroupResource(ModelResource):
    project = fields.ToOneField(ProjectResource, 'project', readonly=True)
    team = fields.ToOneField(TeamResource, 'team', readonly=True)
    message = fields.CharField('message', readonly=True)
    times_seen = fields.IntegerField('times_seen', readonly=True)
    last_seen = fields.DateTimeField('last_seen', readonly=True)
    first_seen = fields.DateTimeField('first_seen', readonly=True)
    latest_event = fields.ToOneField('sentry.api.EventResource', 'latest_event', readonly=True, full=True)
    # events = fields.ToManyField('sentry.api.EventResource', 'event_set')

    class Meta:
        queryset = Group.objects.all()
        resource_name = 'group'
        allowed_methods = ['get']
        fields = ['id']
        authentication = SentryAuthentication()
        authorization = ProjectBasedAuthorization()

    def full_dehydrate(self, bundle):
        # HACK:
        bundle.obj.team = bundle.obj.project.team
        bundle.obj.latest_event = bundle.obj.get_latest_event()
        return super(GroupResource, self).full_dehydrate(bundle)


class EventResource(ModelResource):
    group = fields.ToOneField(GroupResource, 'group', readonly=True)
    project = fields.ToOneField(ProjectResource, 'project', readonly=True)
    message = fields.CharField('message', readonly=True)
    # data = fields.DictField(readonly=True)
    datetime = fields.DateTimeField('datetime', readonly=True)

    class Meta:
        queryset = Event.objects.all()
        resource_name = 'event'
        allowed_methods = ['get']
        fields = ['id']
        authentication = SentryAuthentication()
        authorization = ProjectBasedAuthorization()

    # def dehydrate_data(self, bundle):
    #     return bundle.obj.data


v1_api = Api(api_name='v1')
v1_api.register(TeamResource())
v1_api.register(TeamMemberResource())
v1_api.register(ProjectResource())
v1_api.register(GroupResource())
v1_api.register(EventResource())
