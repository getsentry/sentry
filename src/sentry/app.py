"""
sentry.app
~~~~~~~~~~

:copyright: (c) 2010-2014 by the Sentry Team, see AUTHORS for more details.
:license: BSD, see LICENSE for more details.
"""
from __future__ import absolute_import

from threading import local

from django.conf import settings
from raven.contrib.django.models import client

from sentry.utils import redis
from sentry.utils.imports import import_string
from sentry.utils.locking.backends.redis import RedisLockBackend
from sentry.utils.locking.manager import LockManager


class State(local):
    request = None
    data = {}

env = State()


def get_instance(path, options):
    cls = import_string(path)
    return cls(**options)


# TODO(dcramer): this is getting heavy, we should find a better way to structure
# this
buffer = get_instance(settings.SENTRY_BUFFER, settings.SENTRY_BUFFER_OPTIONS)
digests = get_instance(settings.SENTRY_DIGESTS, settings.SENTRY_DIGESTS_OPTIONS)
quotas = get_instance(settings.SENTRY_QUOTAS, settings.SENTRY_QUOTA_OPTIONS)
nodestore = get_instance(
    settings.SENTRY_NODESTORE, settings.SENTRY_NODESTORE_OPTIONS)
ratelimiter = get_instance(
    settings.SENTRY_RATELIMITER, settings.SENTRY_RATELIMITER_OPTIONS)
search = get_instance(settings.SENTRY_SEARCH, settings.SENTRY_SEARCH_OPTIONS)
tsdb = get_instance(settings.SENTRY_TSDB, settings.SENTRY_TSDB_OPTIONS)
raven = client
locks = LockManager(RedisLockBackend(redis.clusters.get('default')))
