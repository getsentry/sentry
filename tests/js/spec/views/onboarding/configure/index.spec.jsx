import React from 'react';
import {shallow, mount} from 'enzyme';

import {Client} from 'app/api';
import Configure from 'app/views/onboarding/configure';
import SentryTypes from '../../../../../../src/sentry/static/sentry/app/proptypes';

describe('Configure should render correctly', function() {
  let sandbox;
  let stubbedApiRequest;

  beforeEach(function() {
    sandbox = sinon.sandbox.create();
    stubbedApiRequest = sandbox.stub(Client.prototype, 'request');
  });

  afterEach(function() {
    sandbox.restore();
  });

  describe('render()', function() {
    const baseProps = {
      next: () => {},
      params: {
        projectId: 'testProject',
        orgId: 'testOrg',
      },
    };

    it("shouldn't redirect for a found platform", function() {
      let props = {
        ...baseProps,
      };
      props.params.platform = 'node';

      let wrapper = shallow(<Configure {...props} />, {
        context: {organization: {id: '1337', slug: 'testOrg', teams: [['testProject']]}},
        childContextTypes: {organization: SentryTypes.Organization},
      });

      const component = wrapper.instance();

      let handleSubmitStub = sandbox.stub(component, 'redirectToNeutralDocs', () => {});

      wrapper.update();
      expect(wrapper).toMatchSnapshot();
      expect(handleSubmitStub.callCount).toEqual(0);
    });

    it('should redirect to if no matching platform', function() {
      let props = {
        ...baseProps,
      };
      props.params.platform = 'other';

      let handleSubmitStub = sandbox.stub(Configure.prototype, 'redirectToNeutralDocs');

      // üë∫ ‚ö†Ô∏è this is a hack to defeat the method auto binding so we can fully stub the method. It would not be neccessary with es6 class components and it relies on react internals so it's fragile - maxbittker
      const index =
        Configure.prototype.__reactAutoBindPairs.indexOf('redirectToNeutralDocs') + 1;
      Configure.prototype.__reactAutoBindPairs[index] = handleSubmitStub;

      let wrapper = shallow(<Configure {...props} />, {
        context: {
          organization: {
            id: '1337',
            slug: 'testOrg',
            teams: [['testProject']],
          },
        },
      });

      expect(wrapper).toMatchSnapshot();
      expect(handleSubmitStub.callCount).toEqual(1);
    });

    it('should render platform docs', function() {
      let props = {
        ...baseProps,
      };
      props.params.platform = 'node';

      let wrapper = mount(<Configure {...props} />, {
        context: {
          organization: {
            id: '1337',
            slug: 'testOrg',
            projects: [
              {
                name: 'Test Project',
                slug: 'testProject',
                id: 'testProject',
                hasAccess: true,
                teams: [
                  {
                    id: 'coolteam',
                    hasAccess: true,
                  },
                ],
              },
            ],
            teams: [
              {
                id: 'coolteam',
                hasAccess: true,
                projects: [
                  {
                    name: 'Test Project',
                    slug: 'testProject',
                    id: 'testProject',
                  },
                ],
              },
            ],
          },
        },
        childContextTypes: {organization: SentryTypes.Organization},
      });
      expect(wrapper).toMatchSnapshot();
      expect(stubbedApiRequest.callCount).toEqual(5);
    });
  });
});
