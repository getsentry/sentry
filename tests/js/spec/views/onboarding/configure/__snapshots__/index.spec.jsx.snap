// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Configure should render correctly render() should redirect to if no matching platform 1`] = `
<div>
  <div
    className="onboarding-Configure"
  >
    <h2
      style={
        Object {
          "marginBottom": 30,
        }
      }
    >
      Configure your application
    </h2>
    <withProjects
      orgId="testOrg"
      projectId="project-slug"
      style={
        Object {
          "marginBottom": 30,
        }
      }
    >
      <ProjectDocsContext>
        <ProjectInstallPlatform
          linkPath={[Function]}
          params={
            Object {
              "orgId": "testOrg",
              "platform": "other",
              "projectId": "project-slug",
            }
          }
          platformData={
            Object {
              "hack": "actually set by ProjectDocsContext, this object is here to avoid proptypes warnings",
            }
          }
        />
      </ProjectDocsContext>
    </withProjects>
    <Waiting
      hasEvent={false}
      skip={[Function]}
    />
  </div>
</div>
`;

exports[`Configure should render correctly render() should render platform docs 1`] = `
<Configure
  next={[Function]}
  params={
    Object {
      "orgId": "testOrg",
      "platform": "node",
      "projectId": "project-slug",
    }
  }
>
  <div>
    <div
      className="onboarding-Configure"
    >
      <h2
        style={
          Object {
            "marginBottom": 30,
          }
        }
      >
        Configure your application
      </h2>
      <withProjects
        orgId="testOrg"
        projectId="project-slug"
        style={
          Object {
            "marginBottom": 30,
          }
        }
      >
        <withRouter(ProjectContext)
          orgId="testOrg"
          projectId="project-slug"
          projects={
            Array [
              Object {
                "hasAccess": true,
                "id": "testProject",
                "isBookmarked": false,
                "name": "Test Project",
                "slug": "project-slug",
                "teams": Array [
                  Object {
                    "hasAccess": true,
                    "id": "coolid",
                    "slug": "coolteam",
                  },
                ],
              },
            ]
          }
          style={
            Object {
              "marginBottom": 30,
            }
          }
        >
          <ProjectContext
            location={
              Object {
                "query": Object {},
              }
            }
            orgId="testOrg"
            projectId="project-slug"
            projects={
              Array [
                Object {
                  "hasAccess": true,
                  "id": "testProject",
                  "isBookmarked": false,
                  "name": "Test Project",
                  "slug": "project-slug",
                  "teams": Array [
                    Object {
                      "hasAccess": true,
                      "id": "coolid",
                      "slug": "coolteam",
                    },
                  ],
                },
              ]
            }
            router={
              Object {
                "createHref": [MockFunction] {
                  "calls": Array [
                    Array [
                      "/onboarding/testOrg/project-slug/configure/node/",
                    ],
                    Array [
                      "/onboarding/testOrg/project-slug/configure/node/",
                    ],
                  ],
                },
                "go": [MockFunction],
                "goBack": [MockFunction],
                "goForward": [MockFunction],
                "isActive": [MockFunction],
                "listen": [MockFunction],
                "location": Object {
                  "query": Object {},
                },
                "push": [MockFunction],
                "replace": [MockFunction],
                "setRouteLeaveHook": [MockFunction],
              }
            }
            style={
              Object {
                "marginBottom": 30,
              }
            }
          >
            <SideEffect(DocumentTitle)
              title="project-slug"
            >
              <DocumentTitle
                title="project-slug"
              >
                <ProjectDocsContext>
                  <ProjectInstallPlatform
                    linkPath={[Function]}
                    params={
                      Object {
                        "orgId": "testOrg",
                        "platform": "node",
                        "projectId": "project-slug",
                      }
                    }
                    platformData={
                      Object {
                        "dsn": "https://9ed7cdc60:20e868d7b@sentry.io/300733",
                        "dsnPublic": "https://9ed7cdc6581145bcb46044b77bd82aa0@sentry.io/300733",
                        "platforms": Array [
                          Object {
                            "id": "javascript",
                            "integrations": Array [
                              Object {
                                "id": "node",
                                "link": "https://docs.getsentry.com/hosted/clients/csharp/",
                                "name": "node",
                                "type": "language",
                              },
                            ],
                            "name": "js",
                          },
                        ],
                      }
                    }
                  >
                    <div
                      className="install row"
                    >
                      <div
                        className="install-content col-md-10"
                      >
                        <Panel>
                          <Component
                            className="css-10qfvek-Panel e1laxa7d0"
                          >
                            <div
                              className="css-10qfvek-Panel e1laxa7d0"
                            >
                              <PanelHeader
                                hasButtons={true}
                              >
                                <Component
                                  className="css-xm0vov-PanelHeader-getPadding e1p8v8nv0"
                                  hasButtons={true}
                                >
                                  <Flex
                                    align="center"
                                    className="css-xm0vov-PanelHeader-getPadding e1p8v8nv0"
                                    justify="space-between"
                                  >
                                    <Base
                                      align="center"
                                      className="e1p8v8nv0 css-13rknte-PanelHeader-getPadding"
                                      justify="space-between"
                                    >
                                      <div
                                        className="e1p8v8nv0 css-13rknte-PanelHeader-getPadding"
                                        is={null}
                                      >
                                        Configure node
                                        <Button
                                          disabled={false}
                                          external={true}
                                          href="https://docs.getsentry.com/hosted/clients/csharp/"
                                          size="small"
                                        >
                                          <StyledButton
                                            aria-label="Full Documentation"
                                            disabled={false}
                                            external={true}
                                            href="https://docs.getsentry.com/hosted/clients/csharp/"
                                            onClick={[Function]}
                                            role="button"
                                            size="small"
                                          >
                                            <Component
                                              aria-label="Full Documentation"
                                              className="css-dkprmi-StyledButton-getColors e17811v30"
                                              disabled={false}
                                              external={true}
                                              href="https://docs.getsentry.com/hosted/clients/csharp/"
                                              onClick={[Function]}
                                              role="button"
                                              size="small"
                                            >
                                              <ExternalLink
                                                aria-label="Full Documentation"
                                                className="css-dkprmi-StyledButton-getColors e17811v30"
                                                disabled={false}
                                                href="https://docs.getsentry.com/hosted/clients/csharp/"
                                                onClick={[Function]}
                                                rel="noreferrer noopener"
                                                role="button"
                                                size="small"
                                                target="_blank"
                                              >
                                                <a
                                                  aria-label="Full Documentation"
                                                  className="css-dkprmi-StyledButton-getColors e17811v30"
                                                  disabled={false}
                                                  href="https://docs.getsentry.com/hosted/clients/csharp/"
                                                  onClick={[Function]}
                                                  rel="noreferrer noopener"
                                                  role="button"
                                                  size="small"
                                                  target="_blank"
                                                >
                                                  <ButtonLabel
                                                    size="small"
                                                  >
                                                    <Component
                                                      className="css-19bmqb3-ButtonLabel e17811v31"
                                                      size="small"
                                                    >
                                                      <Flex
                                                        align="center"
                                                        className="css-19bmqb3-ButtonLabel e17811v31"
                                                        size="small"
                                                      >
                                                        <Base
                                                          align="center"
                                                          className="e17811v31 css-1eblpyp-ButtonLabel"
                                                          size="small"
                                                        >
                                                          <div
                                                            className="e17811v31 css-1eblpyp-ButtonLabel"
                                                            is={null}
                                                            size="small"
                                                          >
                                                            Full Documentation
                                                          </div>
                                                        </Base>
                                                      </Flex>
                                                    </Component>
                                                  </ButtonLabel>
                                                </a>
                                              </ExternalLink>
                                            </Component>
                                          </StyledButton>
                                        </Button>
                                      </div>
                                    </Base>
                                  </Flex>
                                </Component>
                              </PanelHeader>
                              <PanelBody
                                direction="column"
                                disablePadding={false}
                                flex={false}
                              >
                                <div
                                  className="css-zhfi4l-padding-textStyles"
                                >
                                  <TextBlock>
                                    <Component
                                      className="css-1geyb25-TextBlock ec8ep340"
                                    >
                                      <div
                                        className="css-1geyb25-TextBlock ec8ep340"
                                      >
                                        <span
                                          key="8"
                                        >
                                          <span
                                            key="0"
                                          >
                                            
             This is a quick getting started guide. For in-depth instructions
             on integrating Sentry with 
                                          </span>
                                          <span
                                            key="2"
                                          >
                                            node
                                          </span>
                                          <span
                                            key="3"
                                          >
                                            , view
             
                                          </span>
                                          <a
                                            href="https://docs.getsentry.com/hosted/clients/csharp/"
                                            key="5"
                                          >
                                            <span
                                              key="4"
                                            >
                                              our complete documentation
                                            </span>
                                          </a>
                                          <span
                                            key="6"
                                          >
                                            .
            
                                          </span>
                                        </span>
                                      </div>
                                    </Component>
                                  </TextBlock>
                                  <DocumentationWrapper
                                    dangerouslySetInnerHTML={
                                      Object {
                                        "__html": undefined,
                                      }
                                    }
                                  >
                                    <div
                                      className="css-1y7l4t-DocumentationWrapper e1w9lqpk0"
                                      dangerouslySetInnerHTML={
                                        Object {
                                          "__html": undefined,
                                        }
                                      }
                                    />
                                  </DocumentationWrapper>
                                </div>
                              </PanelBody>
                            </div>
                          </Component>
                        </Panel>
                      </div>
                      <div
                        className="install-sidebar col-md-2"
                      >
                        <LanguageNav
                          active={true}
                          key="javascript"
                          name="js"
                        >
                          <div
                            className="install-language-nav"
                          >
                            <ul
                              className="list-group"
                            >
                              <li
                                className="list-group-item list-group-header"
                              >
                                <a
                                  onClick={[Function]}
                                  style={
                                    Object {
                                      "display": "block",
                                    }
                                  }
                                >
                                  <span
                                    className="pull-right"
                                  >
                                    <span
                                      className="icon-minus"
                                    />
                                  </span>
                                  <strong>
                                    js
                                  </strong>
                                </a>
                              </li>
                              <span
                                style={
                                  Object {
                                    "display": "block",
                                  }
                                }
                              >
                                <Link
                                  className="list-group-item"
                                  key="node"
                                  to="/onboarding/testOrg/project-slug/configure/node/"
                                >
                                  <Link
                                    className="list-group-item"
                                    onlyActiveOnIndex={false}
                                    style={Object {}}
                                    to="/onboarding/testOrg/project-slug/configure/node/"
                                  >
                                    <a
                                      className="list-group-item"
                                      onClick={[Function]}
                                      style={Object {}}
                                    >
                                      node
                                    </a>
                                  </Link>
                                </Link>
                              </span>
                            </ul>
                          </div>
                        </LanguageNav>
                      </div>
                    </div>
                  </ProjectInstallPlatform>
                </ProjectDocsContext>
              </DocumentTitle>
            </SideEffect(DocumentTitle)>
          </ProjectContext>
        </withRouter(ProjectContext)>
      </withProjects>
      <Waiting
        hasEvent={false}
        skip={[Function]}
      >
        <div
          className="awaiting-event"
        >
          <div
            className="pull-right"
          >
            <div
              className="btn btn-primary"
              onClick={[Function]}
            >
              All done!
            </div>
          </div>
          <div
            className="wrap waiting-text"
          >
            <h3
              className="animated-ellipsis"
            >
              Waiting to receive an error
            </h3>
            <div
              className="robot"
            >
              <span
                className="eye"
              />
            </div>
          </div>
        </div>
      </Waiting>
    </div>
  </div>
</Configure>
`;

exports[`Configure should render correctly render() shouldn't redirect for a found platform 1`] = `
<div>
  <div
    className="onboarding-Configure"
  >
    <h2
      style={
        Object {
          "marginBottom": 30,
        }
      }
    >
      Configure your application
    </h2>
    <withProjects
      orgId="testOrg"
      projectId="project-slug"
      style={
        Object {
          "marginBottom": 30,
        }
      }
    >
      <ProjectDocsContext>
        <ProjectInstallPlatform
          linkPath={[Function]}
          params={
            Object {
              "orgId": "testOrg",
              "platform": "node",
              "projectId": "project-slug",
            }
          }
          platformData={
            Object {
              "hack": "actually set by ProjectDocsContext, this object is here to avoid proptypes warnings",
            }
          }
        />
      </ProjectDocsContext>
    </withProjects>
    <Waiting
      hasEvent={false}
      skip={[Function]}
    />
  </div>
</div>
`;
